compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/CompilationRequestManager.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.ballerinalang.compiler;
+
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+import org.wso2.ballerinalang.compiler.util.StaleCompilationRequestException;
+
+/**
+ * This class is responsible for collecting compilation requests that are coming in from various sources and ensuring
+ * that the latest request is maintained.
+ *
+ * @since 2.0.0
+ */
+public class CompilationRequestManager {
+
+    private static final CompilerContext.Key<CompilationRequestManager> COMPILATION_REQUEST_MANAGER_KEY =
+            new CompilerContext.Key<>();
+
+    private CompilationRequest currentCompileRequest;
```

Comment:
This should be marked as volatile to avoid L1, L2, L3 cache line memory visibility issue.